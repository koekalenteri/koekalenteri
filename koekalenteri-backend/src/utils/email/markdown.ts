/* eslint-disable @typescript-eslint/no-explicit-any */
import type { Template } from 'aws-sdk/clients/ses'
import type { Root } from 'mdast'
import type { Plugin } from 'unified'

import remarkBreaks from 'remark-breaks'
import remarkGfm from 'remark-gfm'
import remarkHtml from 'remark-html'
import remarkParse from 'remark-parse'
import { unified } from 'unified'
import { SKIP, visit } from 'unist-util-visit'

import { linkHandler } from './mdast2hast/link'
import { tableHandler } from './mdast2hast/table'
import { linkAsText } from './mdast2text/link'
import { remarkPlainText } from './mdast2text/remarkPlainText'
import { removeTableHead } from './mdast2text/table'

export async function markdownToTemplate(templateName: string, source: string): Promise<Template> {
  const { text, subject } = await markdownToText(source)
  const html = await markdownToHtml(source)

  return {
    TemplateName: templateName,
    SubjectPart: subject,
    TextPart: String(text),
    HtmlPart: String(html),
  }
}

export async function markdownToText(source: string): Promise<{ subject: string; text: string }> {
  let subject = ''
  const extractSubject: Plugin<void[], string, Root> = () => (tree: any) =>
    visit(tree, (node, index, parent) => {
      if (subject === '' && node.type === 'definition' && node.title) {
        subject = node.title
        parent.children.splice(index, 1)
        return [SKIP, index]
      }
    })

  const text = await unified()
    .use(remarkParse)
    .use(extractSubject)
    .use(remarkGfm)
    .use(removeTableHead)
    .use(linkAsText)
    .use(remarkPlainText)
    .process(source)

  return { subject, text: String(text) }
}

export async function markdownToHtml(source: string): Promise<string> {
  const html = await unified()
    .use(remarkParse)
    .use(remarkGfm)
    .use(remarkBreaks)
    .use(remarkHtml, { handlers: { table: tableHandler, link: linkHandler } })
    .process(source)

  return String(html)
}
