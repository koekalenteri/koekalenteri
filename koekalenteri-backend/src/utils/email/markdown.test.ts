import { markdownToHtml, markdownToTemplate, markdownToText } from './markdown'

const testTemplate = '[subject]: # ({{subject}}: {{reg.eventType}} {{eventDate}} {{event.location}})\n\n# {{title}}\n\n{{#unless reg.cancelled ~}}\n[Edit registration]({{link}}/edit)\n[Cancel registration]({{link}}/cancel)\n{{/unless}}\n\ntable header | is removed\n:-- | ----\nEvent          :| {{reg.eventType}} {{eventDate}} {{event.location}} {{#if event.name}}({{event.name}}){{/if}}\nClass          :| {{reg.class}}\nDog            :| {{reg.dog.regNo}} {{reg.dog.name}}\nIdentification :| {{reg.dog.rfid}}\nBreed          :| {{dogBreed}}\nOwne           :| {{reg.owner.name}}\nHometown       :| {{reg.owner.location}}\nEmail          :| {{reg.owner.email}}\nHandler        :| {{reg.handler.name}}\nHometown       :| {{reg.handler.location}}\nEmail          :| {{reg.handler.email}}\nPhone          :| {{reg.handler.phone}}\nTimes          :| {{regDates}}\nReserve        :| {{reserveText}}\n\n**Qualifying results**\n{{#each qualifyingResults}}\n{{this.type}} {{this.result}}, {{this.date}}, {{this.location}}, {{this.judge}}\n{{/each}}\n{{#if reg.notes}}\n\n**Notes**\n{{reg.notes}}\n{{/if}}\n\nYou can view your registration details here: [View registration details]({{link}})\n\nBest regards,\n{{#if event.contactInfo.secretary.name ~}}\n{{event.secretary.name}}\nSecretary\n{{/if ~}}\n{{#if event.contactInfo.secretary.email ~}}\n{{event.secretary.email}}\n{{/if ~}}\n{{#if event.contactInfo.secretary.phone ~}}\np. {{event.secretary.phone}}\n{{/if ~}}\n\n{{#if event.contactInfo.official.name ~}}\n{{event.official.name}}\nChief officer\n{{/if ~}}\n{{#if event.contactInfo.official.email ~}}\n{{event.official.email}}\n{{/if ~}}\n{{#if event.contactInfo.official.phone ~}}\np. {{event.official.phone}}\n{{/if ~}}'

describe('markdown', () => {
  describe('markdownToText', () => {
    it('should return empty string for empty imput', async () => {
      expect(await markdownToText('')).toEqual({subject: '', text: ''})
    })

    it('should convert a template', async () => {
      expect(await markdownToText(testTemplate)).toMatchSnapshot(``)
    })
  })

  describe('markdownToHtml', () => {
    it('should return empty string for empty imput', async () => {
      expect(await markdownToHtml('')).toEqual('')
    })

    it('should convert a template', async () => {
      expect(await markdownToHtml(testTemplate)).toMatchSnapshot(``)
    })
  })

  describe('markdownToTemplate', () => {
    it('should return the template name', async () => {
      expect(await markdownToTemplate('testTemplateName', '')).toEqual({
        TemplateName: 'testTemplateName',
        SubjectPart: '',
        TextPart: '',
        HtmlPart: '',
      })
    })

    it('should return render detect [subject]', async () => {
      expect(await markdownToTemplate('testTemplateName', '[subject]: # (test)\n# content')).toEqual({
        TemplateName: 'testTemplateName',
        SubjectPart: 'test',
        TextPart: 'content\n',
        HtmlPart: '<h1>content</h1>\n',
      })
    })
  })
})
