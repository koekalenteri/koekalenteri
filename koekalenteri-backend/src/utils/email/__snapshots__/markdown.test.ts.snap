// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown markdownToHtml should convert a template 1`] = `
"<h1>{{title}}</h1>
<p>{{#unless reg.cancelled ~}}<br>
<a href="{{link}}/edit">Edit registration</a><br>
<a href="{{link}}/cancel">Cancel registration</a><br>
{{/unless}}</p>
<table>
<tbody>
<tr>
<th align="left">Event          </th>
<td>{{reg.eventType}} {{eventDate}} {{event.location}} {{#if event.name}}({{event.name}}){{/if}}</td>
</tr>
<tr>
<th align="left">Class          </th>
<td>{{reg.class}}</td>
</tr>
<tr>
<th align="left">Dog            </th>
<td>{{reg.dog.regNo}} {{reg.dog.name}}</td>
</tr>
<tr>
<th align="left">Identification </th>
<td>{{reg.dog.rfid}}</td>
</tr>
<tr>
<th align="left">Breed          </th>
<td>{{dogBreed}}</td>
</tr>
<tr>
<th align="left">Owne           </th>
<td>{{reg.owner.name}}</td>
</tr>
<tr>
<th align="left">Hometown       </th>
<td>{{reg.owner.location}}</td>
</tr>
<tr>
<th align="left">Email          </th>
<td>{{reg.owner.email}}</td>
</tr>
<tr>
<th align="left">Handler        </th>
<td>{{reg.handler.name}}</td>
</tr>
<tr>
<th align="left">Hometown       </th>
<td>{{reg.handler.location}}</td>
</tr>
<tr>
<th align="left">Email          </th>
<td>{{reg.handler.email}}</td>
</tr>
<tr>
<th align="left">Phone          </th>
<td>{{reg.handler.phone}}</td>
</tr>
<tr>
<th align="left">Times          </th>
<td>{{regDates}}</td>
</tr>
<tr>
<th align="left">Reserve        </th>
<td>{{reserveText}}</td>
</tr>
</tbody>
</table>
<p><strong>Qualifying results</strong><br>
{{#each qualifyingResults}}<br>
{{this.type}} {{this.result}}, {{this.date}}, {{this.location}}, {{this.judge}}<br>
{{/each}}<br>
{{#if reg.notes}}</p>
<p><strong>Notes</strong><br>
{{reg.notes}}<br>
{{/if}}</p>
<p>You can view your registration details here: <a href="{{link}}">View registration details</a></p>
<p>Best regards,<br>
{{#if event.contactInfo.secretary.name ~}}<br>
{{event.secretary.name}}<br>
Secretary<br>
{{/if ~}}<br>
{{#if event.contactInfo.secretary.email ~}}<br>
{{event.secretary.email}}<br>
{{/if ~}}<br>
{{#if event.contactInfo.secretary.phone ~}}<br>
p. {{event.secretary.phone}}<br>
{{/if ~}}</p>
<p>{{#if event.contactInfo.official.name ~}}<br>
{{event.official.name}}<br>
Chief officer<br>
{{/if ~}}<br>
{{#if event.contactInfo.official.email ~}}<br>
{{event.official.email}}<br>
{{/if ~}}<br>
{{#if event.contactInfo.official.phone ~}}<br>
p. {{event.official.phone}}<br>
{{/if ~}}</p>
"
`;

exports[`markdown markdownToText should convert a template 1`] = `
{
  "subject": "{{subject}}: {{reg.eventType}} {{eventDate}} {{event.location}}",
  "text": "{{title}}

{{#unless reg.cancelled ~}}
Edit registration: {{link}}/edit
Cancel registration: {{link}}/cancel
{{/unless}}

Event          : {{reg.eventType}} {{eventDate}} {{event.location}} {{#if event.name}}({{event.name}}){{/if}}
Class          : {{reg.class}}
Dog            : {{reg.dog.regNo}} {{reg.dog.name}}
Identification : {{reg.dog.rfid}}
Breed          : {{dogBreed}}
Owne           : {{reg.owner.name}}
Hometown       : {{reg.owner.location}}
Email          : {{reg.owner.email}}
Handler        : {{reg.handler.name}}
Hometown       : {{reg.handler.location}}
Email          : {{reg.handler.email}}
Phone          : {{reg.handler.phone}}
Times          : {{regDates}}
Reserve        : {{reserveText}}

Qualifying results
{{#each qualifyingResults}}
{{this.type}} {{this.result}}, {{this.date}}, {{this.location}}, {{this.judge}}
{{/each}}
{{#if reg.notes}}

Notes
{{reg.notes}}
{{/if}}

You can view your registration details here: View registration details: {{link}}

Best regards,
{{#if event.contactInfo.secretary.name ~}}
{{event.secretary.name}}
Secretary
{{/if ~}}
{{#if event.contactInfo.secretary.email ~}}
{{event.secretary.email}}
{{/if ~}}
{{#if event.contactInfo.secretary.phone ~}}
p. {{event.secretary.phone}}
{{/if ~}}

{{#if event.contactInfo.official.name ~}}
{{event.official.name}}
Chief officer
{{/if ~}}
{{#if event.contactInfo.official.email ~}}
{{event.official.email}}
{{/if ~}}
{{#if event.contactInfo.official.phone ~}}
p. {{event.official.phone}}
{{/if ~}}

",
}
`;
